src/Tracer.php

```diff
@@ -0,0 +1,203 @@
+            'fail',
+            $this->methodCode($testMethod),
+            $this->taskId($testMethod),
+            '',
+            $phpUnitMessage,
+        );
+    }
+
+    private function addTestErrored(Errored $event): void
+    {
+        /** @var TestMethod */
+        $testMethod = $event->test();
+
+        $phpUnitMessage = \trim($event->throwable()->asString());
+        $phpUnitMessage = \str_replace(
+            $this->exerciseDir . '/',
+            '',
+            $phpUnitMessage
+        );
+        $phpUnitMessage = $testMethod->nameWithClass() . "\n" . $phpUnitMessage;
+
+        $this->result['tests'][] = new Result(
+            $testMethod->testDox()->prettifiedMethodName(),
+            'error',
+            $this->methodCode($testMethod),
+            $this->taskId($testMethod),
+            '',
+            $phpUnitMessage,
+        );
+    }
+
+    private function addBeforeFirstTestMethodErrored(BeforeFirstTestMethodErrored $event): void
+    {
+        $phpUnitMessage = \trim($event->throwable()->asString());
+        $phpUnitMessage = \str_replace(
+            $this->exerciseDir . '/',
+            '',
+            $phpUnitMessage
+        );
+
+        $this->result['status'] = 'error';
+        $this->result['message'] = $phpUnitMessage;
+    }
+
+    private function addTestOutput(PrintedUnexpectedOutput $event): void
+    {
+        // This must rely on the sequence of events!
+
+        /** @var Result */
+        $lastTest = $this->result['tests'][\array_key_last($this->result['tests'])];
+        $lastTest->setUserOutput($event->output());
+    }
+
+    private function saveResults(): void
+    {
+        foreach ($this->result['tests'] as $result) {
+            if ($result->isFailed() || $result->isErrored()) {
+                $this->result['status'] = 'fail';
+            }
+        }
+
+        \file_put_contents(
+            $this->outFileName,
+            \json_encode($this->result) . "\n",
+            // \json_encode($this->result, JSON_PRETTY_PRINT) . "\n",
+        );
+    }
+
+    private function methodCode(TestMethod $testMethod): string
+    {
+        $reflectionClass = new ReflectionClass($testMethod->className());
+        $reflectionMethod = $reflectionClass->getMethod($testMethod->methodName());
+
+        // Line numbers are 1-based, array index is 0-based.
+        // Reflections start line is the function declaration, end line is
+        // closing curly bracket.
+        // We use PSR-12, which makes line based code extraction problematic
+        // (function parameters may be on multiple lines). But we have 99% of
+        // code starting on second line after function declaration, and the
+        // closing bracket will be on the line after the last code line.
+        $start = $reflectionMethod->getStartLine() - 1 + 2;
+        $end = $reflectionMethod->getEndLine() - 1 - 1;
+
+        $codeLines = \array_filter(
+            \file($reflectionMethod->getFileName()),
```

Comment:
Would it make sense to keep this in a cache?

Reply:
Would be possible, but gave no significant performance gain. Linux caches the file in memory, PHP caches it in memory until `clearstatcache()` is called...